pysmt/typing.py

```diff
@@ -219,11 +220,21 @@ class _FunctionType(PySMTType):
     This class should not be instantiated directly, but the factory
     method FunctionType should be used instead.
     """
+
+    _instances = {}
+
     def __init__(self, return_type, param_types):
-        PySMTType.__init__(self, type_id = 4)
+        PySMTType.__init__(self)
         self._return_type = return_type
-        self._param_types = param_types
+        self._param_types = tuple(param_types)
         self._hash = hash(return_type) + sum(hash(p) for p in param_types)
+        # Note:
+        # To simplify handling of types, we need to treat FunctionType
+        # in a special way. We do so by setting its arity to a
+        # negative value and creating an args that is the concatenation of
+        # return_type and param_types. In this way, we can iterate on args.
+        self.arity = -1
+        self.args = (self._return_type,) + self.param_types
```

Comment:
I clarified the comment there, because it took me some time to recall what I did :)

The idea is that types/sorts create a tree similar to FNode. In FNode we create a distinction btw payload and args. In order to capture the difference between return value and parameters in FunctionType as you suggest, we would need to make the same distinction here. I do not see a big benefit in this. On the contrary, treating all arguments together allows us to iterate over them more easily, and use PySMTType as key in hash functions.

I have no idea why I used self.arity = -1 . I changed it to be consistent with the other cases and be ```len(self.args)```. In fact, we might consider converting this into a property.